source/common/http/header_map_impl.cc

```diff
@@ -443,57 +434,76 @@ void HeaderMapImpl::addCopy(const LowerCaseString& key, uint64_t value) {
     return;
   }
   HeaderString new_key;
-  new_key.setCopy(key.get().c_str(), key.get().size());
+  new_key.setCopy(key.get());
   HeaderString new_value;
   new_value.setInteger(value);
   insertByKey(std::move(new_key), std::move(new_value));
   ASSERT(new_key.empty());   // NOLINT(bugprone-use-after-move)
   ASSERT(new_value.empty()); // NOLINT(bugprone-use-after-move)
+  verifyByteSize();
 }
 
-void HeaderMapImpl::addCopy(const LowerCaseString& key, const std::string& value) {
+void HeaderMapImpl::addCopy(const LowerCaseString& key, absl::string_view value) {
   auto* entry = getExistingInline(key.get());
   if (entry != nullptr) {
     const uint64_t added_size = appendToHeader(entry->value(), value);
     addSize(added_size);
     return;
   }
   HeaderString new_key;
-  new_key.setCopy(key.get().c_str(), key.get().size());
+  new_key.setCopy(key.get());
   HeaderString new_value;
-  new_value.setCopy(value.c_str(), value.size());
+  new_value.setCopy(value);
   insertByKey(std::move(new_key), std::move(new_value));
   ASSERT(new_key.empty());   // NOLINT(bugprone-use-after-move)
   ASSERT(new_value.empty()); // NOLINT(bugprone-use-after-move)
+  // verifyByteSize();
```

Comment:
?

Reply:
Fixed, but quick question about it. I forgot to take this out when testing a protocol integration test's `ManyRequestHeadersTimeout` test failure. The `verifyByteSize()` assertions bring back  slow n^2 verifications in debug mode. So right now the test is skipped in debug mode. What we really want to test is that the time it takes to process one additional header is the same when there's 5 headers or 500, I suppose.

Reply:
Hmm, yeah, I see the problem. @htuch @alyssawilk any thoughts here from a testing infra perspective? It seems not great that we would have tests that only run on optimized builds but I get why this should be the case. 

Perhaps should we run the tests on all builds but just decrease the amount of headers on debug builds with a comment?

/wait-any

Reply:
So this is doing a full header scan on every header change in debug mode to verify the code?  I think that's a bit much even for an ASSERT check - folks regularly run Envoy in non-optimized mode and it's a pretty expensive verification.  I'd be more inclined to do an  to only evaluate this if some global static "should we sanity check" boolean  we override for test is set so we can turn it on for various test paths but it's not on by default

Reply:
FWIW many header modifications appears to be O(n) currently anyway, but constant factors matter too, and obviously the O(1) headers are not.

Anyway I agree with Alyssa on this point that this seems expensive. Maybe we could have it triggered not by a global static, but instead by injecting these sanity checks in TestHeaderMapImpl?

Nevertheless, this is the existing condition for this code, and maybe that could be done in a follow-up?

Reply:
I'm fine with whatever solution you all agree on as long as we have a clear cleanup path.

/wait

Reply:
OK thought about this for a little bit. Injecting in TestHeaderMapImpl will be hard because handling the inline header cases will require a lot of added code. Guarding the code in `verifyByteSize()` with a boolean makes more sense, and it won't look too different if that boolean is a global static or a member variable set to false in HeaderMapImpl and true in TestHeaderMapImpl. Will do the latter.

Reply:
Update: static global was messy, pushed a virtual `verifyByteSize()` that is overridden with the time-consuming verify in `TestHeaderMapImpl`. I added a TODO to remove the empty verification from `HeaderMapImpl`, and for it to exist in `TestHeaderMapImpl` (by making wrapper methods with a verification).

Prod code doesn't perform these sanity checks, only `TestHeaderMapImpl` will.